### DISCUSSION TITLE
Concurrent Full Calculate

### DISCUSSION THREAD

**Discussion by louisskelton** [jamessimone/apex-rollup#524] [Q&A]
Hello again! Incredible that the rollup works across more than 50k records! One thing I've found in my testing, however, is that if a rollup triggers the `RollupFullBatchRecalculator` class and processes the rollup across batches, if a subsequent updates triggers additional instances of that class the resulting calculation can be incorrect. Is that just an inevitability and you just need to be careful in such instances and, for example, only process these rollups on a schedule or manually rather than via a trigger?

**Comment by jamessimone** [comment]
@louisskelton do you mean that you occasionally have two different full recalcs running at the same time (due to two different updates being made)? I can definitely see how that would be an issue, but I want to make sure I understand the "how" here prior to going further.

**Reply by louisskelton** [reply]
Hi @jamessimone - yeah that's it. So, if your parent has 1000s of children and the update is triggered via the apex trigger on one of the children, and then a subsequent update is made, on the same child or another of the children - you end up with 2x `RollupFullBatchRecalculator` running and the calculations on the parent end up being incorrect. I'm guessing it's because of the 2 concurrent async processes (just because that's something I've battled with in the past!) but I guess that could be a red herring!

**Reply by jamessimone** [reply]
That's going to be a tough one to address, to be honest. There's not a lot I can do as a package author to detect for subsequent updates while a full recalc is already running, unfortunately. With that being said, I also want to confirm something else - are we talking about making an update to a parent record that has 25k+ children? That's the current break point beyond which a regular rollup gets transformed to a `RollupFullBatchRecalculator`, but if you have _less_ than 25k children records on the parent record in question, something else must be going on, and it'd be good to understand your exact setup to see if there isn't a bug elsewhere in the logic.
